### PR TITLE
Rework Mimic as Rampant Brand Intelligence

### PR DESCRIPTION
**Pull request by VMSolidus**
# Description

This PR was originally just going to be me fixing the stupid MobMimic bug. Over the course of 5 hours, and in a state of sleep deprived delirium, I somehow expanded this into something resembling the Rampant Brand Intelligence midround event. You're welcome I guess? I'm going to go get some sleep. 

# TODO

- [x] Re-examine this after at least 9 hours of sleep, and vomit after I read my own shitcode
- [x] Make it not shitcode

# Changelog

:cl:
- add: Rampant Brand Intelligence has been spotted infesting NT Stations, be on the lookout for highly aggressive vending machines.

### CODE REVIEW

**Review by DangerRevolution — APPROVED** [review]
very cool

**Comment by VMSolidus** [comment]
I am NOT going to refactor this PR any further than I already have, because I originally set out just to get the damn test fails to stop firing, and already scope creeped the shit out of this PR. So what I have here is done for now, and anything more to expand upon it will have to be a separate PR. I need to get back to doing important things like the Psionic Refactor.

File: Content.Server/Advertise/AdvertiseComponent.cs [hunk 1]
```diff
@@ -3,23 +3,23 @@
...
+    [RegisterComponent]
     public sealed partial class AdvertiseComponent : Component
     {
         /// <summary>
         ///     Minimum time in seconds to wait before saying a new ad, in seconds. Has to be larger than or equal to 1.
         /// </summary>
         [ViewVariables(VVAccess.ReadWrite)]
         [DataField("minWait")]
-        public int MinimumWait { get; private set; } = 8 * 60;
+        public int MinimumWait { get; set; } = 8 * 60;
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Advertise/AdvertiseComponent.cs:14]:
```suggestion
        public int MinimumWait = 8 * 60;
```

File: Content.Server/Advertise/AdvertiseComponent.cs [hunk 2]
```diff
@@ -3,23 +3,23 @@
...
 
         /// <summary>
         ///     Maximum time in seconds to wait before saying a new ad, in seconds. Has to be larger than or equal
         ///     to <see cref="MinimumWait"/>
         /// </summary>
         [ViewVariables(VVAccess.ReadWrite)]
         [DataField("maxWait")]
-        public int MaximumWait { get; private set; } = 10 * 60;
+        public int MaximumWait { get; set; } = 10 * 60;
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Advertise/AdvertiseComponent.cs:22]:
```suggestion
        public int MaximumWait = 10 * 60;
```

File: Content.Server/Advertise/AdvertiseComponent.cs [hunk 3]
```diff
@@ -3,23 +3,23 @@
...
 namespace Content.Server.Advertise
 {
-    [RegisterComponent, Access(typeof(AdvertiseSystem))]
+    [RegisterComponent]
     public sealed partial class AdvertiseComponent : Component
     {
         /// <summary>
         ///     Minimum time in seconds to wait before saying a new ad, in seconds. Has to be larger than or equal to 1.
         /// </summary>
         [ViewVariables(VVAccess.ReadWrite)]
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Advertise/AdvertiseComponent.cs:12]: You should double-check this, but I believe VVRW is added by DataField.

```suggestion
```

File: Content.Server/Advertise/AdvertiseComponent.cs [hunk 4]
```diff
@@ -3,23 +3,23 @@
...
         [DataField("minWait")]
-        public int MinimumWait { get; private set; } = 8 * 60;
+        public int MinimumWait { get; set; } = 8 * 60;
 
         /// <summary>
         ///     Maximum time in seconds to wait before saying a new ad, in seconds. Has to be larger than or equal
         ///     to <see cref="MinimumWait"/>
         /// </summary>
         [ViewVariables(VVAccess.ReadWrite)]
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Advertise/AdvertiseComponent.cs:20]:
```suggestion
```

File: Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs [hunk 5]
```diff
@@ -13,9 +13,33 @@ public sealed partial class MobReplacementRuleComponent : Component
     [DataField]
     public EntProtoId Proto = "MobMimic";
 
+    [DataField("numberToReplace")]
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs:16]:
```suggestion
    [DataField]
```

File: Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs [hunk 6]
```diff
@@ -13,9 +13,33 @@ public sealed partial class MobReplacementRuleComponent : Component
...
+    public int NumberToReplace { get; set; }
+
+    [DataField("announcement")]
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs:19]:
```suggestion
    [DataField]
```

File: Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs [hunk 7]
```diff
@@ -13,9 +13,33 @@ public sealed partial class MobReplacementRuleComponent : Component
...
+    public string Announcement = "station-event-rampant-intelligence-announcement";
+
     /// <summary>
     /// Chance per-entity.
     /// </summary>
     [DataField]
     public float Chance = 0.001f;
+
+    [DataField("doAnnouncement")]
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs:28]:
```suggestion
    [DataField]
```

File: Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs [hunk 8]
```diff
@@ -13,9 +13,33 @@ public sealed partial class MobReplacementRuleComponent : Component
...
+    public bool DoAnnouncement = true;
+
+    [DataField("mimicMeleeDamage")]
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs:31]:
```suggestion
    [DataField]
```

File: Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs [hunk 9]
```diff
@@ -13,9 +13,33 @@ public sealed partial class MobReplacementRuleComponent : Component
...
+    [DataField("mimicMoveSpeed")]
+    public float MimicMoveSpeed = 1f;
+
+    [DataField("mimicAIType")]
+    public string MimicAIType = "SimpleHostileCompound";
+
+    [DataField("mimicSmashGlass")]
+    public bool MimicSmashGlass = true;
+
+    [DataField("vendorModify")]
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs:43]:
```suggestion
    [DataField]
```

File: Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs [hunk 10]
```diff
@@ -13,9 +13,33 @@ public sealed partial class MobReplacementRuleComponent : Component
...
+    [DataField("mimicMeleeDamage")]
+    public float MimicMeleeDamage = 20f;
+
+    [DataField("mimicMoveSpeed")]
+    public float MimicMoveSpeed = 1f;
+
+    [DataField("mimicAIType")]
+    public string MimicAIType = "SimpleHostileCompound";
+
+    [DataField("mimicSmashGlass")]
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs:40]:
```suggestion
    [DataField]
```

File: Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs [hunk 11]
```diff
@@ -13,9 +13,33 @@ public sealed partial class MobReplacementRuleComponent : Component
...
+    [DataField("doAnnouncement")]
+    public bool DoAnnouncement = true;
+
+    [DataField("mimicMeleeDamage")]
+    public float MimicMeleeDamage = 20f;
+
+    [DataField("mimicMoveSpeed")]
+    public float MimicMoveSpeed = 1f;
+
+    [DataField("mimicAIType")]
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs:37]:
```suggestion
    [DataField]
```

File: Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs [hunk 12]
```diff
@@ -13,9 +13,33 @@ public sealed partial class MobReplacementRuleComponent : Component
...
     [DataField]
     public float Chance = 0.001f;
+
+    [DataField("doAnnouncement")]
+    public bool DoAnnouncement = true;
+
+    [DataField("mimicMeleeDamage")]
+    public float MimicMeleeDamage = 20f;
+
+    [DataField("mimicMoveSpeed")]
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/Mimic/MobReplacementRuleComponent.cs:34]:
```suggestion
    [DataField]
```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 13]
```diff
@@ -1,37 +1,172 @@
...
+    [Dependency] private readonly StationSystem _stationSystem = default!;
+    [Dependency] private readonly GameTicker _gameTicker = default!;
+    [Dependency] private readonly ChatSystem _chatSystem = default!;
+    [Dependency] private readonly IPrototypeManager _prototype = default!;
+    [Dependency] private readonly IComponentFactory _componentFactory = default!;
+    [Dependency] private readonly SharedPhysicsSystem _physics = default!;
+    [Dependency] private readonly NpcFactionSystem _npcFaction = default!;
+    [Dependency] private readonly NPCSystem _npc = default!;
+    [Dependency] private readonly TransformSystem _transform = default!;
+    [Dependency] private readonly AdvertiseSystem _advertise = default!;
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:40]:
```suggestion
    [Dependency] private readonly StationSystem _station = default!;
    [Dependency] private readonly GameTicker _gameTicker = default!;
    [Dependency] private readonly ChatSystem _chat = default!;
    [Dependency] private readonly IPrototypeManager _prototype = default!;
    [Dependency] private readonly IComponentFactory _componentFactory = default!;
    [Dependency] private readonly SharedPhysicsSystem _physics = default!;
    [Dependency] private readonly NpcFactionSystem _npcFaction = default!;
    [Dependency] private readonly NPCSystem _npc = default!;
    [Dependency] private readonly TransformSystem _transform = default!;
    [Dependency] private readonly AdvertiseSystem _advertise = default!;

```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 14]
```diff
@@ -1,37 +1,172 @@
...
+        var stations = _gameTicker.GetSpawnableStations();
 
         while (query.MoveNext(out var vendingUid, out _, out var xform))
         {
-            if (!_random.Prob(component.Chance))
+            var ownerStation = _stationSystem.GetOwningStation(vendingUid);
+
+            if (ownerStation == null
+                || ownerStation != stations[0])
+                return;
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:56]: Shouldn't this be a continue?

```suggestion
                continue;
```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 15]
```diff
@@ -1,37 +1,172 @@
...
+
+            // Make sure that we aren't running this on something that is already a mimic
+            if (HasComp<CombatModeComponent>(vendingUid))
                 continue;
 
             spawns.Add((vendingUid, xform.Coordinates));
         }
+        if (spawns == null)
+        {
+            //WTF:THE STATION DOESN'T EXIST, WE MUST BE IN A CONTENT.INTEGRATION TEST! QUICK, PUT A MIMIC AT 0,0!!!
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:66]:
```suggestion
            //WTF THE STATION DOESN'T EXIST! WE MUST BE IN A TEST! QUICK, PUT A MIMIC AT 0,0!!!
```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 16]
```diff
@@ -1,37 +1,172 @@
...
+                || ownerStation != stations[0])
+                return;
+
+            // Make sure that we aren't running this on something that is already a mimic
+            if (HasComp<CombatModeComponent>(vendingUid))
                 continue;
 
             spawns.Add((vendingUid, xform.Coordinates));
         }
+        if (spawns == null)
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:64]:
```suggestion

        if (spawns == null)
```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 17]
```diff
@@ -1,37 +1,172 @@
...
+        {
+            //WTF:THE STATION DOESN'T EXIST, WE MUST BE IN A CONTENT.INTEGRATION TEST! QUICK, PUT A MIMIC AT 0,0!!!
+            Spawn(component.Proto, new EntityCoordinates(uid, new Vector2(0, 0)));
+        }
+        else
+        {
+            //This is intentionally not clamped. If a server host wants to replace every vending machine in the entire station with a mimic, who am I to stop them?
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:71]:
```suggestion
            // This is intentionally not clamped. If a server host wants to replace every vending machine in the entire station with a mimic, who am I to stop them?
```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 18]
```diff
@@ -1,37 +1,172 @@
...
+                k--;
+            }
+        }
+    }
+
+    /// <summary>
+    /// It's like Build a Bear, but MURDER
+    /// </summary>
+    /// <param name="uid"></param>
+    public void BuildAMimicWorkshop(EntityUid uid, MobReplacementRuleComponent component)
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:92]:
```suggestion
    public void BuildAMimic(EntityUid uid, MobReplacementRuleComponent component)
```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 19]
```diff
@@ -1,37 +1,172 @@
...
+    {
+        if (!TryComp<MetaDataComponent>(uid, out var metaData))
+            return;
+
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:96]:
```suggestion
        var metaData = MetaData(uid);
```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 20]
```diff
@@ -1,37 +1,172 @@
...
+                    _chatSystem.DispatchStationAnnouncement(stations[0], Loc.GetString("station-event-rampant-intelligence-announcement"), playDefaultSound: true,
+                        colorOverride: Color.Red, sender: "Central Command");
+
+                k--;
+            }
+        }
+    }
+
+    /// <summary>
+    /// It's like Build a Bear, but MURDER
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:89]:
```suggestion
    /// 	It's like Build a Bear, but MURDER
```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 21]
```diff
@@ -1,37 +1,172 @@
...
+            .Select(name => (name, _componentFactory.GetRegistration(name).Type))
+            .ToList() ?? new List<(string name, Type type)>();
+
+        foreach (var (name, _) in mimicComponents.Except(vendorComponents))
         {
-            var coordinates = entity.Coordinates;
-            Del(entity.Entity);
+            var newComponent = _componentFactory.GetComponent(name);
+            EntityManager.AddComponent(uid, newComponent);
+        }
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:114]:
```suggestion
        }

```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 22]
```diff
@@ -1,37 +1,172 @@
...
+            .Select(name => (name, _componentFactory.GetRegistration(name).Type))
+            .ToList() ?? new List<(string name, Type type)>();
 
-        foreach (var entity in spawns)
+        var mimicComponents = mimicProto?.Components.Keys
+            .Where(n => n != "Transform" && n != "MetaData")
+            .Select(name => (name, _componentFactory.GetRegistration(name).Type))
+            .ToList() ?? new List<(string name, Type type)>();
+
+        foreach (var (name, _) in mimicComponents.Except(vendorComponents))
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:110]: Isn't this a `List`?

```suggestion
        foreach (var name in mimicComponents.Except(vendorComponents))
```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 23]
```diff
@@ -1,37 +1,172 @@
...
+            _transform.Unanchor(uid, xform);
 
-            Spawn(component.Proto, coordinates);
+        SetupMimicNPC(uid, component);
+
+        if (TryComp<AdvertiseComponent>(uid, out var vendor)
+            && component.VendorModify)
+        {
+            SetupMimicVendor(uid, component, vendor);
         }
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:125]:
```suggestion
            SetupMimicVendor(uid, component, vendor);
```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 24]
```diff
@@ -1,37 +1,172 @@
...
     }
+    /// <summary>
+    /// This handles getting the entity ready to be a hostile NPC
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:128]:
```suggestion
    /// 	This handles getting the entity ready to be a hostile NPC
```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 25]
```diff
@@ -1,37 +1,172 @@
...
+        melee.Damage = dspec;
+
+        var movementSpeed = EnsureComp<MovementSpeedModifierComponent>(uid);
+        (movementSpeed.BaseSprintSpeed, movementSpeed.BaseWalkSpeed) = (component.MimicMoveSpeed, component.MimicMoveSpeed);
+
+        var htn = EnsureComp<HTNComponent>(uid);
+        htn.RootTask = new HTNCompoundTask() { Task = component.MimicAIType };
+        htn.Blackboard.SetValue(NPCBlackboard.NavSmash, component.MimicSmashGlass);
+        _npc.WakeNPC(uid, htn);
+    }
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:155]:
```suggestion
    }

```

File: Content.Server/Antag/MobReplacementRuleSystem.cs [hunk 26]
```diff
@@ -1,37 +1,172 @@
...
+    /// <summary>
+    /// Handling specific interactions with vending machines
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Antag/MobReplacementRuleSystem.cs:157]:
```suggestion
    /// Handling specific interactions with vending machines

```

File: Resources/Locale/en-US/station-events/events/rempant-brand-intelligence.ftl [hunk 27]
```diff
@@ -0,0 +1 @@
+station-event-rampant-intelligence-announcement= Rampant brand intelligence has been detected board your station. Please inspect any vendors for aggressive marketting tactics, and reboot them if necessary.
```
Comment (DEATHB4DEFEAT) [inline Resources/Locale/en-US/station-events/events/rempant-brand-intelligence.ftl:1]:
```suggestion
station-event-rampant-intelligence-announcement = Rampant brand intelligence has been detected board your station. Please inspect any vendors for aggressive marketing tactics, and reboot them if necessary.
```

**Review by DEATHB4DEFEAT — CHANGES_REQUESTED** [review]

File: Content.Server/Antag/MobReplacementRuleSystem.cs (hunk 18 above)
Comment (VMSolidus) [reply Content.Server/Antag/MobReplacementRuleSystem.cs:92]: No.

Comment (DEATHB4DEFEAT) [reply Content.Server/Antag/MobReplacementRuleSystem.cs:94]: Why?

Comment (VMSolidus) [reply Content.Server/Antag/MobReplacementRuleSystem.cs:94]: I wanted to leave a silly joke in the code. I mean if it must be removed I'll remove it. :)

**Comment by DangerRevolution** [comment]
re-requested @DEATHB4DEFEAT review because it was addressed

**Comment by DangerRevolution** [comment]
test fail seems real; didn't look too much

**Review by DEATHB4DEFEAT — APPROVED** [review]

**Comment by VMSolidus** [comment]
Test fail is legitimate, and I genuinely don't know what the test fail means or what it's referring to. Don't merge this please.
Also I just.. Completely forgot this PR even existed in the first place.

**Comment by DEATHB4DEFEAT** [comment]
> Test fail is legitimate, and I genuinely don't know what the test fail means or what it's referring to. Don't merge this please. Also I just.. Completely forgot this PR even existed in the first place.

Merged master to see if it'll succeed.

**Comment by VMSolidus** [comment]
image
Has to do with how the Pick function works when selecting a vending machine to replace.

**Pull request merged**